### PR TITLE
Use Bazelisk launcher for Bazel.

### PR DESCRIPTION
**Pull request by PiotrSikora**
Signed-off-by: Piotr Sikora <piotrsikora@google.com>

### CODE REVIEW

**Comment by PiotrSikora** [comment]
Note: someone needs to push new images, since I don't have the permission to do it.

**Review by howardjohn — APPROVED** [review]

File: scripts/tools/linux-install-bazel [hunk 1]
```diff
@@ -8,31 +8,32 @@ fi
...
+# Use Bazelisk launcher for Bazel (automatically downloads required version of Bazel).
+BAZELISK_VERSION="v1.0"
+BAZELISK_BASE_URL="https://github.com/bazelbuild/bazelisk/releases/download"
+BAZELISK_BIN="bazelisk-linux-amd64"
+BAZELISK_URL="${BAZELISK_BASE_URL}/${BAZELISK_VERSION}/${BAZELISK_BIN}"
+BAZELISK_DIRECTORY="${TOOLS_DIR}/bazelisk"
 
 function update_bazel() {
-  local bazel_current="$(bazel version | grep -oh '[0-9]*\.[0-9]*\.[0-9]*')" \
+  local bazel_current="$(bazel version | grep -oh 'Bazelisk version: v[0-9]*\.[0-9]*')" \
```
Comment (clarketm) [inline scripts/tools/linux-install-bazel:19]: This should also be `bazelisk` correct?
```suggestion
  local bazel_current="$(bazelisk version | grep -oh 'Bazelisk version: v[0-9]*\.[0-9]*')" \
```

Comment (PiotrSikora) [reply scripts/tools/linux-install-bazel:19]: No, it's installed as `/usr/local/bin/bazel`.

**Review by fejta — COMMENTED** [review]
Is this for developers or test containers?

**Comment by PiotrSikora** [comment]
> Is this for developers or test containers?

No, this installs Bazelisk on build images (in place of Bazel).

**Comment by fejta** [comment]
Will this be used in CI? If so I'm concerned that this will result in every job downloading the one of a couple versions of bazel over and over again. Is there a way we can prepopulate the image with the versions of bazel we expect?

File: scripts/tools/linux-install-bazel (hunk 1 above)
Comment (clarketm) [reply scripts/tools/linux-install-bazel:19]: Okay, I see.

**Comment by PiotrSikora** [comment]
@fejta yes, it's going to be used and downloaded on every build in CI... unless we can store/restore cache in Prow? See https://github.com/istio/proxy/pull/2371 for how it's stored/restored in CircleCI.

This is unfortunate side-effect of this change, but the productivity drain that results from merging new images each time new Bazel is needed are worth it (...at least, for me).

**Comment by fejta** [comment]
> Prow doesn't provide a runtime. Your image is your runtime. So you need to make sure your image has already downloaded the right bazel versions.
>
> From https://github.com/bazelbuild/bazelisk#how-does-bazelisk-know-which-bazel-version-to-run-and-where-to-get-it-from it looks one way to do this would be to `export USE_BAZEL_VERSION=0.28.1 bazelisk version` (and do that for a couple versions) in the Dockerfile.
> 
> That way we will create images that have 0.28.1 (etc) of bazel preinstalled

Yes, we could preinstall the few versions we need for the Istio Proxy releases we still maintain, but that will get outdated pretty soon (i.e. on the next release).

**Comment by PiotrSikora** [comment]
@fejta added few know versions that we need.

File: docker/prowbazel/Dockerfile [hunk 2]
```diff
@@ -33,6 +33,11 @@ WORKDIR ${HOME}
 # code and running bazel in different dir than the calling script.
 ENV TEST_TMPDIR /home/bootstrap/.cache/bazel
 
+# Move preloaded Bazel binaries to bootstrap's cache.
+RUN mkdir -p /home/bootstrap/.cache/
```
Comment (fejta) [inline docker/prowbazel/Dockerfile]: Each `RUN` statement creates a new docker layer. I'd recommend trying to combine these as much as possible to reduce the number of layers we need to download and extract:

Running multiple commands in one layer (at least combine these three commands into one):
* https://github.com/kubernetes/test-infra/blob/1972ae8f994d61b88f7a84c9947aa88bcb3aae18/images/bootstrap/Dockerfile#L30-L52

Adding multiple env vars into one layer (if you want to do a more aggressive refactor):
* https://github.com/kubernetes/test-infra/blob/1972ae8f994d61b88f7a84c9947aa88bcb3aae18/images/bootstrap/Dockerfile#L21-L22

Comment (PiotrSikora) [reply docker/prowbazel/Dockerfile]: Done & I don't.

**Review by fejta — APPROVED** [review]

**Comment by PiotrSikora** [comment]
@fejta could you `prowbazel:0.5.9` image to `gcr.io/istio-testing/`?

**Comment by howardjohn** [comment]
I built and pushed

**Pull request merged**